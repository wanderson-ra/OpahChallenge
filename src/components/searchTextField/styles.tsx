import { View, TextInput, Platform } from "react-native";
import { isIphoneX } from "react-native-iphone-x-helper";
import Ripple from "react-native-material-ripple";
import { widthPercentageToDP as width, heightPercentageToDP as height } from "react-native-responsive-screen";
import Icon from "react-native-vector-icons/MaterialCommunityIcons";

import styled from "styled-components/native";

import { metrics } from "src/configurations/metrics";

export const Container = styled.View`
    height: ${isIphoneX() ? height(5) : height(6)}px;
`;

export const TextInputContainer = styled(View)`
    width: ${width(95)}px;
    height: ${isIphoneX() ? height(5) : height(6)}px;
    background-color: ${(props): string => props.theme.textInput.backgroundColor};
    border-color: ${(props): string => props.theme.textInput.borderColor};
    border-width: 0.5px;
    border-radius: ${metrics.textInput.borderRadius}px;
    padding-left: ${width(3)}px;
    padding-right: ${width(3)}px;
    justify-content: space-evenly;
    align-items: center;
    flex-direction: row;
`;

export const IconTextInput = styled(Icon)`
    opacity: 0.6;
`;

export const SearchTextInput = styled(TextInput)`
    margin-left: ${width(2)}px;
    color: ${(props): string => props.theme.textInput.color};
    font-size: ${metrics.font.medium}px;
    width: ${width(60)}px;
    height: ${isIphoneX() ? height(5) : height(6)}px;
    padding-top: ${Platform.select({
        android: height(2),
        ios: 0,
    })}px;
`;

export const ClearButton = styled(Ripple)`
    height: ${metrics.icon.large + 15}px;
    width: ${metrics.icon.large + 15}px;
    border-radius: ${(metrics.icon.large + 15) / 2}px;
    opacity: 0.6;
    justify-content: center;
    align-items: center;
`;

export const ClearIcon = styled(Icon)``;
