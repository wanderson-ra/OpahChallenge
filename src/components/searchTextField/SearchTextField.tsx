import React, { useState, useEffect, useCallback } from "react";
import { StyleProp, ViewStyle, TextStyle } from "react-native";

import { useTheme } from "styled-components/native";

import { metrics } from "src/configurations/metrics";

import { Container, TextInputContainer, IconTextInput, SearchTextInput, ClearButton, ClearIcon } from "./styles";

export interface ValidatedTextInputProps {
    value: string;
    placeholder?: string;
    placeholderTextColor?: string;
    selectionColor?: string;
    onChangeText: (text: string) => void;
    containerStyle?: StyleProp<ViewStyle>;
    textInputStyle?: StyleProp<TextStyle>;
}

export const SearchTextField: React.FC<ValidatedTextInputProps> = (validatedTextInputProps) => {
    const theme = useTheme();

    const {
        value,
        containerStyle,
        textInputStyle,
        onChangeText,
        placeholder,
        placeholderTextColor,
        selectionColor,
    } = validatedTextInputProps;

    const [isEditingEmail, setIsEditingEmail] = useState(false);

    useEffect(() => {
        setIsEditingEmail(value.length > 0);
    }, [value]);

    const handlerOnChangeText = useCallback(
        (text: string): void => {
            onChangeText(text);
        },
        [onChangeText]
    );

    const handlerClearTextInput = useCallback((): void => {
        handlerOnChangeText("");
    }, [handlerOnChangeText]);

    return (
        <Container>
            <TextInputContainer style={containerStyle}>
                <IconTextInput size={metrics.icon.large} name={"magnify"} color={theme.icon.default} />
                <SearchTextInput
                    autoCapitalize={"none"}
                    testID={"input"}
                    selectionColor={selectionColor ? selectionColor : theme.textInput.placeholder}
                    value={value}
                    keyboardType={"default"}
                    returnKeyType={"search"}
                    onChangeText={handlerOnChangeText}
                    style={textInputStyle}
                    placeholder={placeholder}
                    placeholderTextColor={placeholderTextColor ? placeholderTextColor : theme.textInput.placeholder}
                />

                <ClearButton
                    testID={"clearButton"}
                    rippleColor={isEditingEmail ? theme.ripple : "transparent"}
                    onPress={(): void => handlerClearTextInput()}
                >
                    {isEditingEmail ? (
                        <ClearIcon
                            size={metrics.icon.medium}
                            name={"close-circle-outline"}
                            color={theme.icon.default}
                        />
                    ) : null}
                </ClearButton>
            </TextInputContainer>
        </Container>
    );
};
