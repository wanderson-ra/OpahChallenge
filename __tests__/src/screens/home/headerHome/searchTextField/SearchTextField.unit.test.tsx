import React from "react";
import Ripple from "react-native-material-ripple";

import { render, cleanup, fireEvent } from "@testing-library/react-native";
import { ThemeProvider } from "styled-components/native";

import colorTheme from "../../../../../../src/configurations/themes/theme";

import { SearchTextField } from "../../../../../../src/screens/home/components/headerHome/searchTextField/SearchTextField";

describe("Tests of SearchTextField", () => {
    beforeEach(() => {
        cleanup();
    });

    it("Text snapshot", async () => {
        const value = "";
        const placeholder = "anyPlaceholder";
        const onChangeText = jest.fn();

        const searchTextField = render(
            <ThemeProvider theme={colorTheme.light}>
                <SearchTextField onChangeText={onChangeText} placeholder={placeholder} value={value} />
            </ThemeProvider>
        );

        expect(searchTextField.toJSON()).toMatchSnapshot();
    });

    it("Text with text changed and clear text", async () => {
        const value = "";
        const placeholder = "anyPlaceholder";
        const onChangeText = jest.fn();

        const newValue = "anyNewValue";

        const searchTextField = render(
            <ThemeProvider theme={colorTheme.light}>
                <SearchTextField onChangeText={onChangeText} placeholder={placeholder} value={value} />
            </ThemeProvider>
        );

        const input = searchTextField.getByTestId("input");
        fireEvent(input, "onChangeText", newValue);
        expect(onChangeText).toHaveBeenCalledWith(newValue);

        const clearButton = searchTextField.UNSAFE_getByType(Ripple);
        fireEvent(clearButton, "onPress");
        expect(onChangeText).toHaveBeenCalledWith("");
    });
});
