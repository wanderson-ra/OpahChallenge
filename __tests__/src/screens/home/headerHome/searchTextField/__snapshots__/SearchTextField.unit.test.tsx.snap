// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests of SearchTextField Text snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "height": 53.5,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#FFFFFF40",
          "borderColor": "#757575",
          "borderRadius": 13.5,
          "borderWidth": 0.5,
          "flexDirection": "row",
          "height": 53.5,
          "justifyContent": "space-evenly",
          "paddingLeft": 22.5,
          "paddingRight": 22.5,
          "width": 450,
        },
      ]
    }
  >
    <Text
      allowFontScaling={false}
      style={
        Array [
          Object {
            "color": "#757575",
            "fontSize": 45.5,
          },
          Array [
            Object {
              "opacity": 0.6,
            },
          ],
          Object {
            "fontFamily": "Material Design Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      Û∞çâ
    </Text>
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      keyboardType="default"
      onChangeText={[Function]}
      placeholder="anyPlaceholder"
      placeholderTextColor="#757575"
      rejectResponderTermination={true}
      returnKeyType="search"
      selectionColor="#757575"
      style={
        Array [
          Object {
            "color": "#757575",
            "fontSize": 30.5,
            "height": 53.5,
            "marginLeft": 15,
            "paddingTop": 0,
            "width": 300,
          },
        ]
      }
      testID="input"
      underlineColorAndroid="transparent"
      value=""
    />
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onLayout={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      pointerEvents="box-only"
      style={
        Object {
          "alignItems": "center",
          "borderRadius": 30.25,
          "height": 60.5,
          "justifyContent": "center",
          "opacity": 0.6,
          "width": 60.5,
        }
      }
      testID="clearButton"
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "bottom": 0,
              "left": 0,
              "overflow": "hidden",
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
            Object {
              "borderRadius": 0,
            },
          ]
        }
      />
    </View>
  </View>
</View>
`;
